addon/-private/system/store.js

```diff
@@ -2107,13 +2107,40 @@ Store = Service.extend({
         throw new EmberError(`No model was found for '${modelName}'`);
       }
 
-      // interopt with the future
-      let klass = getOwner(this).factoryFor ? factory.class : factory;
+      // interop with the future
+      const hasFactoryFor = getOwner(this).factoryFor;
+      let klass = hasFactoryFor ? factory.class : factory;
 
       assert(`'${inspect(klass)}' does not appear to be an ember-data model`, klass.isModel);
 
-      // TODO: deprecate this
-      klass.modelName = klass.modelName || modelName;
+      // TODO: deprecate klass.modelName
+      /*
+        In test environments it is common to re-use the same base model in multiple
+        tests and register it each time. Since that seems convenient and good
+        to continue allowing this ignores the case where modelName is set but is
+         the same.
+       */
+      if (klass.modelName && klass.modelName !== modelName) {
+        const Message =
+          `Ember Data found a Model for ${modelName} but the class ` +
+          `already had the modelName ${klass.modelName}.\n\n` +
+          'This likely means that you are re-exporting a Model.\n\n' +
+          `Replace:\n\texport { default } from './${klass.modelName}';\nwith:\n\t` +
+          `import Model from './${klass.modelName}';\n\n\texport default Model.extend();`;
+
+        if (hasFactoryFor) {
+          klass = factory.class = klass.extend();
```

Comment:
factoryFor isn’t supposed to allow that, I thought we were using proxies to error in debug for that.